src/main/java/org/radarcns/managementportal/MpClient.java

```diff
@@ -0,0 +1,312 @@
+ * limitations under the License.
+ */
+
+/**
+ * Client to interact with the RADAR Management Portal.
+ */
+public class MpClient {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(MpClient.class);
+
+    private ArrayList<Subject> subjects;
+    private ServletContext context;
+    private boolean isSubjectsInitialised = false;
+
+    /**
+     * @param context {@link ServletContext} useful to retrieve shared {@link OkHttpClient} and
+     *      {@code access token}.
+     * @throws IllegalStateException in case the object cannot be created
+     * @see org.radarcns.listener.managementportal.listener.HttpClientListener
+     * @see org.radarcns.listener.managementportal.listener.TokenManagerListener
+     */
+    public MpClient(ServletContext context) {
+        Objects.requireNonNull(context);
+
+        // TODO Use try catch and initialize all
+
+        this.context = context;
+
+    }
+
+
+    /**
+     * Retrieves all {@link Subject} from the already computed list of subjects
+     * using {@link ArrayList} of {@link Subject} else it calls a method for retrieving
+     * the subjects from MP.
+     * @return {@link ArrayList} of {@link Subject} if a subject is found
+     */
+    public ArrayList<Subject> getSubjects() {
+        if ( isSubjectsInitialised ) {
+            return subjects;
+        } else {
+            try {
+                return getAllSubjects(context);
+            } catch (MalformedURLException exc) {
+                LOGGER.error(exc.getMessage());
+            } catch (URISyntaxException exc) {
+                LOGGER.error(exc.getMessage());
+            } catch (IllegalStateException exc) {
+                LOGGER.error("Error : ", exc.fillInStackTrace());
+            }
+        }
+        LOGGER.warn("Subjects cannot be retrieved");
+        return null;
+    }
+
+    /**
+     * Retrieves all {@link Subject} from Management Portal using {@link ServletContext} entity.
+     * @param context {@link ServletContext} that has been used to authenticate token
+     * @return {@link ArrayList} of {@link Subject} retrieved from the Management Portal
+     * @throws MalformedURLException,URISyntaxException in case the subjects cannot be retrieved
+     */
+    private ArrayList<Subject> getAllSubjects(ServletContext context) throws
+            MalformedURLException, URISyntaxException {
+
+        Request request = getBuilder(Properties.getSubjectEndPoint(), context).get().build();
+
+        ArrayList<Subject> allSubjects;
+
+        try (Response response = HttpClientListener.getClient(context)
+                .newCall(request).execute()) {
+            if (response.isSuccessful()) {
+
+                String jsonData = response.body().string();
+                allSubjects = Subject.getAllSubjectsFromJson(jsonData);
+                LOGGER.info("Retrieved Subjects from MP.");
+                isSubjectsInitialised = true;
+                response.close();
```

Comment:
response is closed by the try-with-resources block, no need to do it again

Reply:
this is also true for all following `response.close()` calls

Reply:
Okay updating now.